test/torchaudio_unittest/functional/batch_consistency_test.py

```diff
@@ -217,3 +217,18 @@ def test_compute_kaldi_pitch(self):
         batch = waveform.view(self.batch_size, n_channels, waveform.size(-1))
         self.assert_batch_consistency(
             F.compute_kaldi_pitch, batch, sample_rate=sample_rate)
+
+    def test_lfilter(self):
+        signal_length = 2048
+        torch.manual_seed(2434)
+        x = torch.randn(self.batch_size, signal_length)
+        a = torch.rand(self.batch_size, 3)
+        b = torch.rand(self.batch_size, 3)
+
+        batchwise_output = F.lfilter(x, a, b)
+        itemwise_output = torch.stack([
+            F.lfilter(x[i], a[i], b[i])
+            for i in range(self.batch_size)
+        ])
+
+        self.assertEqual(batchwise_output, itemwise_output)
```

Comment:
Can we use `self.assert_batch_consistency` helper method? It handles `dtype`/`device` as well.

Reply:
`self.assert_batch_consistency` seems assume only need to take batch on input, but in our case, `a_coeffs` and `b_coeffs` should also be in batch as well.

Reply:
Does that mean it is applying different filters to samples in batch? I thought the same set of filters are applied to each sample in batch, so one can change the batch size without changing `a_coeffs` and `b_coeffs`.

Reply:
Sounds like we need to do 2 tests. I will add another one that use `self.assert_batch_consistency`.

Reply:
And can you clarify that in the above case, where you said "`a_coeffs` and `b_coeffs` should also be in batch as well", the number of filter bank happens to be same as the batch size, but that's not requirement?

So my understanding/expectation is that when input batch is the shape of `[batch_size, sequence_length]`, `a_coeffs` and `b_coeffs` can take any shape of `[filter_dim, number_of_filters]`, without being constrained on the input shape.

And if I understand correctly, here your test is testing that filter banks produces the same result regardless they are applied separately or together, in that correct?

Reply:
> And can you clarify that in the above case, where you said "`a_coeffs` and `b_coeffs` should also be in batch as well", the number of filter bank happens to be same as the batch size, but that's not requirement?
> 
> So my understanding/expectation is that when input batch is the shape of `[batch_size, sequence_length]`, `a_coeffs` and `b_coeffs` can take any shape of `[filter_dim, number_of_filters]`, without being constrained on the input shape.
> 
In this case, when input is a 2D batch of signals, `a_coeffs` and `b_coeffs` should be in shape of `[batch_size, filter_order + 1]` or just `[filter_order + 1]`. The first one means that the number of filters is equal to batch_size, and each signal is applied with different filter; the second is just one filter apply on all signals.

The case that filter shape will not be constrainted, is when the shape of input is  `[..., 1, sequence_length]`. Then  `a_coeffs` and `b_coeffs` can be in any shape of 2D matrix `[number_of_filters, filter_order + 1]`, , the output shape will be `[..., number_of_filters, sequence_length]`.

> And if I understand correctly, here your test is testing that filter banks produces the same result regardless they are applied separately or together, in that correct?

Yes, that's correct.

Reply:
Wait, I think the batch behavior we want to test is actually the coefficients, not the input. 
So we might need to change the test, with `a_coeffs` and `b_coeffs` is the batch input, `waveform` is the parameter.

Reply:
> > So my understanding/expectation is that when input batch is the shape of `[batch_size, sequence_length]`, `a_coeffs` and `b_coeffs` can take any shape of `[filter_dim, number_of_filters]`, without being constrained on the input shape.
> 
> In this case, when input is a 2D batch of signals, `a_coeffs` and `b_coeffs` should be in shape of `[batch_size, filter_order + 1]` or just `[filter_order + 1]`. The first one means that the number of filters is equal to batch_size, and each signal is applied with different filter; the second is just one filter apply on all signals.
> 
> The case that filter shape will not be constrainted, is when the shape of input is `[..., 1, sequence_length]`. Then `a_coeffs` and `b_coeffs` can be in any shape of 2D matrix `[number_of_filters, filter_order + 1]`, the output shape will be `[..., number_of_filters, sequence_length]`. It means each signal is filtered by a shared set of filters.
> 
> > And if I understand correctly, here your test is testing that filter banks produces the same result regardless they are applied separately or together, in that correct?
> 
> Yes, that's correct.

@yoyololicon 

Can you help me clarify with the understanding of the shape semantics here?
 - When the input batch is 2D, the first dimension is interpreted as channel, thus the number of filters has to match the number of channels.
 - When an input signal has multiple channels, then filters have to have multiple channels.
Are these correct?

Reply:
@mthrok 

> Can you help me clarify with the understanding of the shape semantics here?
> 
> * When the input batch is 2D, the first dimension is interpreted as channel, thus the number of filters has to match the number of channels.
> * When an input signal has multiple channels, then filters have to have multiple channels.
>   Are these correct?

If you want to apply multiple filters at once, these are correct; if there is only one filter, it will fall back to original behavior.
The shape semantics I proposed actually follows [pytorch conventions](https://pytorch.org/docs/stable/notes/broadcasting.html) except the last dimension, which is time or filter order.